### PR TITLE
Suggest valid resource pools in vic-machine

### LINKED ISSUE #1004: vic-machine suggest valid resource pools
If the user provides an invalid value for `--compute-resource`, echo a list of valid values when displaying the error message then exit

### PR DESCRIPTION
**Pull request by andrewtchin**
Suggests valid resource pools for the `--compute-resource` CLI arg

#### Invalid user/pass
```
INFO[2016-06-14T14:54:51-07:00] ### Installing VCH ####                      
INFO[2016-06-14T14:54:51-07:00] Generating certificate/key pair - private key in ./chin-vic-24-key.pem 
Error:
Could not determine datacenter from specified --compute-resource path: asdf
Failed to log in to https://skinner.eng.vmware.com/sdk: ServerFaultCode: Cannot complete login due to an incorrect user name or password.
Failed to connect. Verify --target, --user, and --password
Exiting...
```

#### Invalid target
```
INFO[2016-06-14T14:55:48-07:00] ### Installing VCH ####                      
INFO[2016-06-14T14:55:48-07:00] Generating certificate/key pair - private key in ./chin-vic-24-key.pem 
Error:
Could not determine datacenter from specified --compute-resource path: asdf
Failed to connect to https://askinner.eng.vmware.com/sdk: Post https://askinner.eng.vmware.com/sdk: dial tcp: lookup askinner.eng.vmware.com on 127.0.1.1:53: no such host
Failed to connect. Verify --target, --user, and --password
Exiting...
```

#### Nonexistent compute-resource
Warn that datacenter is invalid
```
INFO[2016-06-15T18:50:24-07:00] ### Installing VCH ####                      
INFO[2016-06-15T18:50:24-07:00] Generating certificate/key pair - private key in ./chin-vic-26-key.pem 
INFO[2016-06-15T18:50:25-07:00] Validating supplied configuration            
INFO[2016-06-15T18:50:26-07:00] Invalid datacenter. Unable to suggest valid --compute-resource 
Error:
Failure finding host (): please specify a datacenter
Failure finding pool (): please specify a datacenter
Exiting...
```

#### Nonexistent compute-resource with path that validates
Warn that datacenter is invalid
```
INFO[2016-06-15T18:50:07-07:00] ### Installing VCH ####                      
INFO[2016-06-15T18:50:07-07:00] Generating certificate/key pair - private key in ./chin-vic-26-key.pem 
INFO[2016-06-15T18:50:07-07:00] Validating supplied configuration            
INFO[2016-06-15T18:50:09-07:00] Invalid datacenter. Unable to suggest valid --compute-resource 
Error:
Failure finding dc (asdf): datacenter 'asdf' not found
Failure finding cluster (/asdf/host/A): compute resource '/asdf/host/A' not found
Failure finding ds (/asdf/datastore/Datastore): datastore '/asdf/datastore/Datastore' not found
Failure finding host (): please specify a datacenter
Failure finding pool (/asdf/host/A): resource pool '/asdf/host/A' not found
Exiting...
```

#### Invalid cluster
Suggests all resource pools on the vCenter
```
INFO[2016-06-15T18:57:26-07:00] ### Installing VCH ####                      
INFO[2016-06-15T18:57:26-07:00] Generating certificate/key pair - private key in ./chin-vic-26-key.pem 
INFO[2016-06-15T18:57:26-07:00] Validating supplied configuration            
INFO[2016-06-15T18:57:31-07:00] Suggesting valid values for --compute-resource 
INFO[2016-06-15T18:57:32-07:00] Failed to find resource pool in the provided path. Showing all resource pools. 
INFO[2016-06-15T18:57:33-07:00]   /Datacenter/host/Austin/Resources          
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/zach   
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/chin   
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/josh-pool 
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/sophie-vhc 
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/ian_is_awesome 
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/loc-pool 
INFO[2016-06-15T18:57:34-07:00]     /Datacenter/host/Austin/Resources/line   
INFO[2016-06-15T18:57:34-07:00]   /Datacenter/host/JoshCluster/Resources     
INFO[2016-06-15T18:57:35-07:00]     /Datacenter/host/JoshCluster/Resources/ResourcePool 
Error:
Failure finding cluster (/Datacenter/host/A): compute resource '/Datacenter/host/A' not found
Failure finding pool (/Datacenter/host/A/Resources/asdf): resource pool '/Datacenter/host/A/Resources/asdf' not found
Exiting...
```

#### Valid cluster, invalid RP
Suggests resource pools in the cluster and the cluster path
```
INFO[2016-06-15T18:50:43-07:00] ### Installing VCH ####                      
INFO[2016-06-15T18:50:43-07:00] Generating certificate/key pair - private key in ./chin-vic-26-key.pem 
INFO[2016-06-15T18:50:43-07:00] Validating supplied configuration            
INFO[2016-06-15T18:50:49-07:00] Suggesting valid values for --compute-resource 
INFO[2016-06-15T18:50:50-07:00]   /Datacenter/host/Austin/Resources          
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/zach   
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/chin   
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/josh-pool 
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/sophie-vhc 
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/ian_is_awesome 
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/loc-pool 
INFO[2016-06-15T18:50:51-07:00]     /Datacenter/host/Austin/Resources/line   
Error:
Failure finding pool (/Datacenter/host/Austin/Resources/asdf): resource pool '/Datacenter/host/Austin/Resources/asdf' not found
Exiting...

```

#### Valid RP
```
INFO[2016-06-14T15:00:02-07:00] ### Installing VCH ####                      
INFO[2016-06-14T15:00:02-07:00] Generating certificate/key pair - private key in ./chin-vic-25-key.pem 
INFO[2016-06-14T15:00:02-07:00] Validating supplied configuration            
INFO[2016-06-14T15:00:06-07:00] Network hasan-test exists                    
INFO[2016-06-14T15:00:07-07:00] Creating a Resource Pool                     
INFO[2016-06-14T15:00:07-07:00] Creating appliance on target                 
INFO[2016-06-14T15:00:09-07:00] Uploading images for container               
INFO[2016-06-14T15:00:09-07:00] 	/home/chin/go/src/github.com/vmware/vic/bin/bootstrap.iso 
INFO[2016-06-14T15:00:09-07:00] 	/home/chin/go/src/github.com/vmware/vic/bin/appliance.iso 
INFO[2016-06-14T15:00:26-07:00] Waiting for IP information                   
INFO[2016-06-14T15:00:45-07:00] Initialization of appliance successful       
INFO[2016-06-14T15:00:45-07:00]                                              
INFO[2016-06-14T15:00:45-07:00] SSH to appliance (default=root:password)     
INFO[2016-06-14T15:00:45-07:00] ssh root@10.17.109.101                       
INFO[2016-06-14T15:00:45-07:00]                                              
INFO[2016-06-14T15:00:45-07:00] Log server:                                  
INFO[2016-06-14T15:00:45-07:00] https://10.17.109.101:2378                   
INFO[2016-06-14T15:00:45-07:00]                                              
INFO[2016-06-14T15:00:45-07:00] Connect to docker:                           
INFO[2016-06-14T15:00:45-07:00] docker -H 10.17.109.101:2376 --tls --tlscert='./chin-vic-25-cert.pem' --tlskey='./chin-vic-25-key.pem' info 
INFO[2016-06-14T15:00:45-07:00] Installer completed successfully...   
```

Fixes #1004

### CODE REVIEW

**Comment by dougm** [comment]
Suggested values should include the ComputeResource itself (standalone hosts/clusters), via Finder.ComputeResourceList.  Not sure we even want to suggest the child ResourcePools?  The child pools in the output example above are the pools created by other VCH instances on the system, right?

And since the datacenter arg has already been validated, the suggestions should probably be relative to that.  For example: 'Austin' and 'JoshCluster'

**Comment by emlin** [comment]
@dougm Now the compute-resources is using absolute path, and suppose only accept resource pool. We'll get datacenter, cluster from this parameter.

Filter out other VCH resource pool is a good idea. 
But right now, we still do not have any special attributes set for VCH resource pool. It's doable to check if there is VM in that resource pool with same name, and with VCH guestinfo, but if other VCH creation failed, or in the mean time of other VCH deletion, this check might be wrong.

File: cmd/vic-machine/validate/validator.go [hunk 1]
```diff
@@ -58,8 +58,18 @@ func (v *Validator) Validate(input *data.Data) (*metadata.VirtualContainerHostCo
 	vchConfig.ApplianceSize.Memory.Limit = int64(input.MemoryMB)
 	vchConfig.Name = input.DisplayName
 
+	var errs []string
 	if err := v.ParseComputeResourcePath(input.ComputeResourcePath); err != nil {
-		return nil, err
+		errs = append(errs, err.Error())
```
Comment (emlin) [inline cmd/vic-machine/validate/validator.go]: Do not call getConnection here. Instead, you can bypass the later parameter settings from derived datacenter name and cluster name.
Actually, those code should be refactored to set only when the relied parameter is not "".
Then it's possible to suggest value for them in the future.

File: cmd/vic-machine/validate/validator.go [hunk 2]
```diff
@@ -58,8 +58,18 @@ func (v *Validator) Validate(input *data.Data) (*metadata.VirtualContainerHostCo
...
+		connected, err := v.getConnection(input) // Connect so we can suggest valid values
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		if connected == true {
+			log.Info("Suggesting valid values for compute-resource")
+			v.suggestAllResourcePools()
```
Comment (emlin) [inline cmd/vic-machine/validate/validator.go]: With the above changes, you can move the suggest methods to getResources method.

File: cmd/vic-machine/validate/validator.go [hunk 3]
```diff
@@ -130,25 +140,25 @@ func (v *Validator) getConnection(input *data.Data) error {
...
-	}
+	v.Session = session.NewSession(sessionconfig)
+
 	if _, err = v.Session.Connect(v.Context); err != nil {
-		return err
+		return false, errors.Errorf("%s\nFailed to connect. Verify --target, --user, and --password", err.Error())
 	}
-	return nil
+
+	if _, err = v.Session.Populate(v.Context); err != nil {
```
Comment (emlin) [inline cmd/vic-machine/validate/validator.go]: If you move the suggest method to getResources method, you don't need to populate session here. This one should not be used.

File: cmd/vic-machine/validate/validator.go [hunk 4]
```diff
@@ -222,6 +273,7 @@ func (v *Validator) getResources(ctx context.Context) error {
 		v.Session.Pool, err = finder.ResourcePoolOrDefault(ctx, v.ResourcePoolPath)
 		if err != nil {
 			errs = append(errs, fmt.Sprintf("Failure finding pool (%s): %s", v.ResourcePoolPath, err.Error()))
+			v.suggestComputeResource()
```
Comment (emlin) [inline cmd/vic-machine/validate/validator.go]: Have the suggest method in this method. And verify all the scenarios you described in the PR.

**Comment by dougm** [comment]
@emlin ok, but there is still the main issue that top-level ResourcePool(s) are not included in the list of suggestions - which would be the most likely candidates.

File: cmd/vic-machine/validate/validator.go [hunk 5]
```diff
@@ -174,6 +184,47 @@ func (v *Validator) validateConfiguration(input *data.Data, vchConfig *metadata.
...
+		for _, c := range clusters {
+			numPools += v.listResourcePools(c)
+		}
+	}
+	return numPools
+}
+
+func (v *Validator) listResourcePools(c *object.ComputeResource) int {
+	children := fmt.Sprintf("%s/Resources/*", c.InventoryPath)
+	pools, _ := v.Session.Finder.ResourcePoolList(v.Context, children)
```
Comment (emlin) [inline cmd/vic-machine/validate/validator.go]: Here should add root resource pool, %s/Resources, as one candidate as well.

**Comment by emlin** [comment]
@dougm agree. Had comments for that, thanks.

File: cmd/vic-machine/validate/validator.go (hunk 5 above)
Comment (andrewtchin) [inline cmd/vic-machine/validate/validator.go]: Fixed

**Comment by andrewtchin** [comment]
Updated to include top level resource pools. Example output updated.

**Comment by dougm** [comment]
lgtm

**Comment by emlin** [comment]
LGTM 👍 
But still cannot suggest for invalid datacenter. Would suggest to open one bug to track that issue.

**Pull request merged**